holoviews/operation/element.py

```diff
@@ -700,6 +700,14 @@ class histogram(Operation):
       Used for setting a common style for histograms in a HoloMap or AdjointLayout.""")
 
     def _process(self, element, key=None):
+
+        bad_bins_msg = 'bins only accept numerical data.'
+        if (isinstance(self.p.bins, np.ndarray) and
+           (self.p.bins.dtype.kind in 'SU' or self.bins.dtype.kind == 'O' and not isdatetime(self.bins))):
+            raise TypeError(bad_bins_msg)
+        elif isinstance(self.p.bins, (list, tuple)) and any(not is_number(bin) for bin in self.p.bins):
+            raise TypeError(bad_bins_msg)
```

Comment:
Given all these boolean expressions, it would be good to pull them out onto separate lines for clarity.